### PR TITLE
Add support for vips in the image analyzer

### LINKED ISSUE #41940: ActiveStorage image analyzer does not handle exif rotation for mirrored images
https://github.com/rails/rails/blob/ec5c9a8fcd39335bf79540366370c3dbeb297d13/activestorage/lib/active_storage/analyzer/image_analyzer.rb#L51-L53

should include TopRight and BottomLeft as well

Screen Shot 2021-04-12 at 7 03 50 PM

I switched to use a Vips analyzer which works correctly https://github.com/rails/rails/issues/41940

### LINKED ISSUE #41941: ActiveStorage ImageAnalyzer does not support Vips
Since ActiveStorage can use Vips for everything else, it doesn't make sense that mini_magick and ImageMagick are required to get the image size.

### PR DESCRIPTION
**Pull request by brenogazzola**
### Summary
Fixes: #41940 
Fixes: #41941

Adds support for using vips as an image analyzer in active storage.

### Other Information
This is pretty much my first non trivial PR, so advice on how to improve this is welcome. Just trying to get the ball running on this one, since I'd like to use this too.

### CODE REVIEW

File: activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb
Comment (ghiculescu) [inline activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb:12]: It looks like this method is the same in the `ImageMagick` and `Vips` classes. Could you DRY it up?

File: activestorage/test/models/variant_test.rb
Comment (ghiculescu) [inline activestorage/test/models/variant_test.rb:195]: I'm probably missing it - what's the reasoning behind this change?

File: activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb
Comment (brenogazzola) [reply activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb:12]: Will do.

File: activestorage/test/models/variant_test.rb
Comment (brenogazzola) [reply activestorage/test/models/variant_test.rb:195]: When I originally opened this PR all my vips tests were failing in the CI, even though they passed on the dev box. I got the help of @matthewd and we discovered that the CI environment did not have VIPS installed, and therefore the tests for vips were not running. He asked me to ensure that all vips tests raised errors in the CI, but were allowed to be skipped in contributor's machines to make development easier.

Comment (ghiculescu) [reply activestorage/test/models/variant_test.rb:195]: If the tests are passing now, I assume that means vips has been installed in CI?

Comment (brenogazzola) [reply activestorage/test/models/variant_test.rb:209]: Yes. https://github.com/rails/buildkite-config/commit/2c9273742d285784020ac8ee8ffd099c9abb6615

Comment (zzak) [reply activestorage/test/models/variant_test.rb:209]: Just re-approved the CI to run, let us know if theres any other blockers! :bow:

**Comment by brenogazzola** [comment]
Changes made. Tests passed. There is an error in "Guides" but it seems it is unrelated, and Github says the build passed? There's 2 workflows awaiting approval. What should I do in this case?

**Comment by zzak** [comment]
@brenogazzola In the case where workflows are awaiting approval, one of us has to approve to run the CI 😢 Running again but I think based on your last comment we should be good!

File: activestorage/lib/active_storage/engine.rb
Comment (ghiculescu) [inline activestorage/lib/active_storage/engine.rb:29]: This looks like a breaking change. What happens on this branch if you try to use `ActiveStorage::Analyzer::ImageAnalyzer` directly?

Comment (brenogazzola) [reply activestorage/lib/active_storage/engine.rb:29]: Yes, it is. I didn't think about that. I can return my previous code, but it was looking pretty ugly:

```
def metadata
  if ActiveStorage.variant_processor == :vips
    Analyzer::ImageAnalyzer::Vips.new(@blob).metadata
  else
    Analyzer::ImageAnalyzer::ImageMagick.new(@blob).metadata
  end     
end
```

It also does not fit the pattern that was set with the PDF previewers, where each lib is a different previewer in the array, instead of having a generic `PDFPreviewer` that then chooses the correct class based on the libs available:
```
config.active_storage.previewers = [ 
  ActiveStorage::Previewer::PopplerPDFPreviewer, 
  ActiveStorage::Previewer::MuPDFPreviewer, 
  ActiveStorage::Previewer::VideoPreviewer 
]
```

I'm not sure where to go from here: breaking change, different patterns, or changing the previewers to have a generic `PDFPreviewer` that then decides between `mu` and `poppler`.

Comment (brenogazzola) [reply activestorage/lib/active_storage/engine.rb:29]: I've also just noticed I am not respecting the naming pattern and should probably have named the analyzers `VipsImageAnalyzer` and `ImageMagickImageAnalyzer` so they match the way the previewers are named.

Comment (brenogazzola) [reply activestorage/lib/active_storage/engine.rb:29]: Or I could return the original `ImageAnalyzer` and just add the VipsVersion ahead of it:
```
config.active_storage.analyzers = [ 
  ActiveStorage::Analyzer::VipsImageAnalyzer, 
  ActiveStorage::Analyzer::ImageAnalyzer, 
  ActiveStorage::Analyzer::VideoAnalyzer 
]
```

This would not be a breaking change and it would more or less follow the way the previewers are set. I'd just need another class to place the common `metadata` method so it stayed DRY.

Comment (ghiculescu) [reply activestorage/lib/active_storage/engine.rb:29]: I think the approach here is good, and if users have been doing things like `config.active_storage.analyzers << MyAnalyzer` then it shouldn't cause problems. The risk I can think of is code like `config.active_storage.analyzers.insert(config.active_storage.analyzers.index(ActiveStorage::Analyzer::ImageAnalyzer) + 1, MyAnalyzer)` breaking. Which seems like a bad pattern anyway.

Can you please update the docs: https://github.com/rails/rails/blob/main/guides/source/configuring.md#configuring-active-storage

Comment (brenogazzola) [reply activestorage/lib/active_storage/engine.rb:29]: Done. I've also updated the comments in the code. Unfortunately, the CI failed on the same place as last time (the bug report templates I think)

Comment (ghiculescu) [reply activestorage/lib/active_storage/engine.rb:29]: you might need @zzak to approve it again

Comment (zzak) [reply activestorage/lib/active_storage/engine.rb:29]: @brenogazzola @ghiculescu Approved! :pray:

**Comment by zzak** [comment]
@brenogazzola Can you rebase and squash please? :bow:

**Comment by brenogazzola** [comment]
@zzak Done. Though it will conflict again with my https://github.com/rails/rails/pull/42392 since both updated the `CHANGELOG` and the same line in `CONTRIBUTING`. :sweat_smile:

**Comment by zzak** [comment]
@brenogazzola Now that #42392 was merged hopefully this will be easier to rebase, sorry I missed that until now :bow:

**Comment by brenogazzola** [comment]
@zzak It's fine. It's my own fault for opening multiple PRs for the same feature in the Rails at the same time. I'll keep rebasing as needed.

**Comment by brenogazzola** [comment]
Also, done.

**Review by zzak — APPROVED** [review]

**Review by pixeltrix — APPROVED** [review]
@brenogazzola if can you rebase this, then I'll merge 👍🏻

**Comment by brenogazzola** [comment]
@pixeltrix Rebased. 👍

**Comment by pixeltrix** [comment]
@brenogazzola thanks! 👍🏻

**Comment by wilg** [comment]
Thanks for fixing my issues @brenogazzola!

**Comment by sedubois** [comment]
It appears that with this change away from ImageMagick, the analyzer does not store the image's `width` and `height` in the metadata any more, it only saves `analyzed: true`. Is that intended? Was that information used for something to begin with?

This test used to pass before I started to migrate to Rails 7.0.0.rc3:

```rb
  test "poster is attached" do
    content = contents(:blog_post_1)

    assert content.poster.attached?

    blob = content.poster.blob.tap(&:analyze)

    assert_equal "cover.jpg", blob.filename.to_s
    assert_equal "image/jpeg", blob.content_type
    assert_equal({ identified: true, width: 1600, height: 900, analyzed: true }, # <-- Test fails here
                 blob.metadata.symbolize_keys)
    assert_equal 211486, blob.byte_size
    assert_equal "gDjkvuzYz1oUlbJ9hjbd5w==", blob.checksum
    assert_equal "public", blob.service_name
  end
```

**Comment by brenogazzola** [comment]
Active Storage has tests specifically to ensure that `width` and `height` are being extract for both ImageMagick and Vips. You can see them in the files changed by this PR.

While upgrading, did you notice that Vips has become the new default in Rails 7.0? It's possible that you don't have Vips installed and this is causing the analyzer to fail.

**Comment by sedubois** [comment]
I am using Vips already in my 6.1.4 app. I will try to dig deeper to see why those width/height are disappearing in my test.

**Comment by brenogazzola** [comment]
Is it happening to all images or only some of them? There are 3 `rescue` points in the analyzer that would cause it to skip setting the dimensions:

1 - `ruby-vips` is not installed
2 - The image is not valid
3 - There was a generic error in vips.

All of them generate log when hit. Try going to `test.rb` and changing `log_level` to `info` before running the test again to see which rescue it's hitting.

**Comment by sedubois** [comment]
Thanks for your suggestions. After upgrading my local vips binary (`brew upgrade vips`), the test passes, so I suppose that was the issue!

**Comment by ahmedsayedabdelsalam** [comment]
i thinks this is a breaking change 

in 6.0.1 anyone extending the image analyzer and uses the read_image method and upgrading to rails 7 will get undefined method read_image

**Pull request merged**
